packages/flutter/lib/src/services/restoration.dart

```diff
@@ -0,0 +1,888 @@
+    });
+    _childrenToAdd.clear();
+  }
+
+  void _removeChildData(RestorationBucket child) {
+    assert(child != null);
+    assert(child._parent == this);
+    if (_claimedChildren.remove(child.id) == child) {
+      _rawChildren.remove(child.id.value);
+      if (_rawChildren.isEmpty) {
+        _rawData.remove(_childrenMapKey);
+      }
+      _manager?.scheduleSerialization();
+      return;
+    }
+    _childrenToAdd[child.id]?.remove(child);
+    if (_childrenToAdd[child.id]?.isEmpty == true) {
+      _childrenToAdd.remove(child.id);
+    }
+  }
+
+  void _addChildData(RestorationBucket child) {
+    assert(child != null);
+    assert(child._parent == this);
+    if (_claimedChildren.containsKey(child.id)) {
+      // Delay addition until the end of the frame in the hopes that the current
+      // owner of the child with the same id will have given up that child by
+      // then.
+      _childrenToAdd.putIfAbsent(child.id, () => <RestorationBucket>{}).add(child);
+      _manager.scheduleSerialization(finalizer: _finalize);
+      return;
+    }
+    _finalizeAddChildData(child);
+    _manager.scheduleSerialization();
+  }
+
+  void _finalizeAddChildData(RestorationBucket child) {
+    assert(_claimedChildren[child.id] == null);
+    assert(_rawChildren[child.id.value] == null);
+    _claimedChildren[child.id] = child;
+    _rawChildren[child.id.value] = child._rawData;
+  }
+
+  void _visitChildren(_BucketVisitor visitor, {bool concurrentModification = false}) {
+    Iterable<RestorationBucket> children = _claimedChildren.values
+        .followedBy(_childrenToAdd.values.expand((Set<RestorationBucket> set) => set));
+    if (concurrentModification) {
+      children = children.toList(growable: false);
+    }
+    children.forEach(visitor);
+  }
+
+  // Bucket management
+
+  /// Changes the [RestorationId] under which the bucket is stored in its parent
+  /// to `newId`.
+  ///
+  /// No-op if the bucket is already stored under the provided id.
+  void rename(RestorationId newId) {
+    assert(newId != null);
+    assert(_parent != null);
+    if (newId == id) {
+      return;
+    }
+    _parent._removeChildData(this);
+    _id = newId;
+    _parent._addChildData(this);
+  }
+
+  /// Deletes the bucket and all the data stored in it from the bucket
+  /// hierarchy.
+  ///
+  /// After [dispose] has been called, the data stored in this bucket and its
+  /// children are no longer part of the app's restoration data. The data
+  /// originally stored in the bucket will not be available again when the
+  /// application is restored to this state in the future. It is up to the
+  /// owners of the children to either move them (via [adoptChild]) to a new
+  /// parent that is still part of the bucket hierarchy or to [dispose] of them
+  /// as well.
+  ///
+  /// This method must only be called by the object's owner.
+  @override
+  void dispose() {
+    _parent?._removeChildData(this);
+    _parent = null;
+    _manager = null;
```

Comment:
Can we assert in any methods above that the object hasn't been disposed?

Should we just fold performDecommision into this?

Reply:
The assert is a great idea. I will do this next week (left a TODO for now, so I will not forget until next week).

> Should we just fold performDecommision into this?

See other answer why decommission is more then just a dispose.